brain/morphology.cpp

```diff
@@ -0,0 +1,222 @@
+
+/* Copyright (c) 2013-2015, EPFL/Blue Brain Project
+ *                          Juan Hernando <jhernando@fi.upm.es>
+ *
+ * This file is part of Brion <https://github.com/BlueBrain/Brion>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "morphology.h"
+
+#include <brion/morphology.h>
+#include <brion/circuit.h>
+
+#include <lunchbox/log.h>
+
+#include <boost/foreach.hpp>
+
+#include <bitset>
+
+namespace brain
+{
+
+namespace
+{
+    void nullDeleter( const brion::Morphology* ) {}
+}
+
+class Morphology::Impl
+{
+public:
+    const brion::Vector4fsPtr points;
+    const brion::Vector2isPtr sections;
+    const brion::SectionTypesPtr types;
+    const brion::Vector2isPtr apicals;
+    Vector3f somaPosition;
+
+    Impl( const brion::Morphology& morphology,
+          const MorphologyRepairStage stage )
+        : points( morphology.readPoints( stage ))
+        , sections( morphology.readSections( stage ))
+        , types( morphology.readSectionTypes( ))
+        , apicals( morphology.readApicals( ))
+        , somaPosition( Vector3f::ZERO ) // the soma is assumed to be
+                                         // centered at the 0, 0, 0
+    {
+    }
+
+    uint32_ts getSectionIDs( const SectionTypes& requestedTypes ) const
+    {
+        std::bitset< brion::SECTION_APICAL_DENDRITE > bits;
+        BOOST_FOREACH( const SectionType type, requestedTypes )
+            bits[size_t( type )] = true;
+
+        uint32_ts result;
+        for( uint32_t i = 0; i != types->size(); ++i )
+        {
+            const SectionType type = ( *types )[i];
+            if( type > brion::SECTION_APICAL_DENDRITE + 1 )
```

Comment:
You're right, it's wrong. It should have been type == ...
I thought about it but there's already a SECTION_UNDEFINED. I'd prefer to add SECTION_ALL (internal) as in other enums.